src/azure-cli/azure/cli/command_modules/acs/agentpool_decorator.py

```diff
@@ -0,0 +1,218 @@
+                )
+            )
+        self.cmd = cmd
+        self.raw_param = raw_parameters
+        self.models = models
+        self.decorator_mode = decorator_mode
+        self.intermediates = dict()
+        self.agentpool = None
+
+    def attach_agentpool(self, agentpool: AgentPool) -> None:
+        """Attach the AgentPool object to the context.
+
+        The `agentpool` object is only allowed to be attached once, and attaching again will raise a CLIInternalError.
+
+        :return: None
+        """
+        if self.agentpool is None:
+            self.agentpool = agentpool
+        else:
+            msg = "the same" if self.agentpool == agentpool else "different"
+            raise CLIInternalError(
+                "Attempting to attach the `agentpool` object again, the two objects are {}.".format(
+                    msg
+                )
+            )
+
+    def get_resource_group_name(self) -> str:
+        """Obtain the value of resource_group_name.
+
+        Note: resource_group_name will not be decorated into the `agentpool` object.
+
+        This is a required parameter and its value should be provided by user explicitly.
+
+        :return: string
+        """
+        # read the original value passed by the command
+        resource_group_name = self.raw_param.get("resource_group_name")
+
+        # this parameter does not need dynamic completion
+        # this parameter does not need validation
+        return resource_group_name
+
+    def get_cluster_name(self) -> str:
+        """Obtain the value of cluster_name.
+
+        Note: cluster_name will not be decorated into the `agentpool` object.
+
+        This is a required parameter and its value should be provided by user explicitly.
+
+        :return: string
+        """
+        # read the original value passed by the command
+        cluster_name = self.raw_param.get("cluster_name")
+
+        # this parameter does not need dynamic completion
+        # this parameter does not need validation
+        return cluster_name
+
+    def get_nodepool_name(self) -> str:
+        """Obtain the value of nodepool_name.
+
+        Note: SDK performs the following validation {'required': True, 'pattern': r'^[a-z][a-z0-9]{0,11}$'}.
+
+        This is a required parameter and its value should be provided by user explicitly.
+
+        :return: string
+        """
+        # read the original value passed by the command
+        nodepool_name = self.raw_param.get("nodepool_name")
+        # try to read the property value corresponding to the parameter from the `agentpool` object
+        if self.agentpool and self.agentpool.name is not None:
+            nodepool_name = self.agentpool.name
+
+        # this parameter does not need dynamic completion
+        # validation
+        instances = cf_agent_pools.list(self.get_resource_group_name, self.get_cluster_name)
+        for agentpool_profile in instances:
+            if agentpool_profile.name == nodepool_name:
```

Comment:
is our agent pool case-sensitive?

Reply:
Yes and no. Only lowercase is allowed. See [doc](https://docs.microsoft.com/en-us/azure/aks/troubleshooting#what-naming-restrictions-are-enforced-for-aks-resources-and-parameters).